### PR TITLE
`<valarray>`: Cleanups and length checking for binary operations

### PR DESCRIPTION
**Pull request by frederick-vs-ja**
Follows up #3887.

This PR
- fuses `_Tidy_init` and `_Assign_rv` into callers/default member initializers, with some simplification, and
- adds `noexcept` to some internal functions, which was lacked in the previous PR, and
- requires that two `valarray`s to have the same length when passed to binary operations.

`atan2` and `pow` currently have no preconditions, I've filed LWG-3964 for them.

### CODE REVIEW

**Review by AlexGuteniev — APPROVED** [review]

**Comment by AlexGuteniev** [comment]
> Many operations call `size()`, but it seems OK to directly use `_Mysize`. Should we change to directly use `_Mysize`?

This should improve unoptimized codegen and debugging experience.

For optimized codegen, my suggestion is that even if you use the data member directly, keep local variable, like `const size_t _Size = size()` => `const size_t _Size = _Mysize` and keep using size. The reason is that the compiler can always exclude aliasing of a local variable, but not a data member, if valarray data has also same type as underlying `size_t` type.

**Comment by StephanTLavavej** [comment]
@frederick-vs-ja
> Many operations call `size()`, but it seems OK to directly use `_Mysize`. Should we change to directly use `_Mysize`?

Calling `size()` makes it easier to see the remaining occurrences that could modify `_Mysize`, but using `_Mysize` is friendlier for debug codegen. I'd say I have a weak preference for `_Mysize`.

@AlexGuteniev 
> For optimized codegen, my suggestion is that even if you use the data member directly, keep local variable, like `const size_t _Size = size()` => `const size_t _Size = _Mysize` and keep using size. The reason is that the compiler can always exclude aliasing of a local variable, but not a data member, if valarray data has also same type as underlying `size_t` type.

I'd worry that that makes the code harder to understand, because it introduces more state that we have to worry about. I'd accept it if benchmarking showed significant improvement, but otherwise I'd advise against it as a pre-emptive micro-optimization.

**Comment by AlexGuteniev** [comment]
> I'd worry that that makes the code harder to understand, because it introduces more state that we have to worry about. I'd accept it if benchmarking showed significant improvement, but otherwise I'd advise against it as a pre-emptive micro-optimization.

That can be order of magnitude, really.

Here's a quick example where the case with local size has AVX2 after some checks, and the case with member size doesn't:
```C++
#include <stdlib.h>

struct Va
{

    void add1(const Va& other);
    void add2(const Va& other);

    size_t* data;
    size_t  size;
};


void Va::add1(const Va& other) {
    for (size_t i = 0; i < size; ++i) {
        data[i] += other.data[i];
    }
}

void Va::add2(const Va& other){
    for (size_t i = 0, s = size; i < s; ++i) {
        data[i] += other.data[i];
    }
}
```
https://godbolt.org/z/6j4T4one5

Do I need to provide a benchmark with measuremets, or looking into codegen is enough?

**Comment by AlexGuteniev** [comment]
Note that unlike other compilers, MSVC does not take advantage of the Strict Aliasing Rule, so will pessimize the member `size_t` usage for all types of elements, not just alias-able with `size_t` !

**Comment by StephanTLavavej** [comment]
I think the codegen is sufficiently convincing, thanks. I see now that `valarray` is special because it actually has its own loops that operate on elements, unlike `vector` which cares about its elements far less. Should we file a tracking issue to audit usage of `size()`?

File: stl/inc/valarray [hunk 1]
```diff
@@ -316,6 +298,9 @@ public:
     }
 
     valarray& operator/=(const valarray& _Right) {
```
Comment (StephanTLavavej) [inline stl/inc/valarray:300]: Missed `operator*=(const valarray& _Right)` above.

**Review by StephanTLavavej — APPROVED** [review]

**Comment by AlexGuteniev** [comment]
> I see now that `valarray` is special because it actually has its own loops that operate on elements, unlike `vector` which cares about its elements far less.

Yes, with `vector` you would either use STL algorithms, which would cache `end()` a it is a parameter, or a loop in user's code (which is out of STL control and up to the user).

> Should we file a tracking issue to audit usage of `size()`?

I'd like that instead of doing this in this PR.

**Comment by StephanTLavavej** [comment]
I've filed #3915. I strongly agree that such a large, semi-mechanical change should be kept separate from this PR's different large-ish changes.

**Comment by StephanTLavavej** [comment]
I've pushed a merge with `main` to resolve a bunch of trivial adjacent-edit conflicts with #3968.

**Review by StephanTLavavej — APPROVED** [review]

**Review by CaseyCarter — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
I'm mirroring this to the MSVC-internal repo - please notify me if any further changes are pushed.

**Comment by StephanTLavavej** [comment]
Thanks for improving these debug checks! :heavy_check_mark: :white_check_mark: :ballot_box_with_check:

**Pull request merged**